### PR TITLE
DM-33994: Add threadpoolctl (RFC-835).

### PR DESCRIPTION
**Pull request by ktlim**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by ktlim** [comment]
@conda-forge-admin, please rerender

**Review by timj — APPROVED** [review]

**Comment by erykoff** [comment]
We need @beckermr's help on pinning the compiler versions so they don't get updated without a specific plan/tests on our side.

**Comment by beckermr** [comment]
@conda-forge-admin rerender

**Comment by beckermr** [comment]
@conda-forge-admin rerender

**Review by erykoff — APPROVED** [review]
Looks good!  And can you tell us where the `conda_build_config.yaml` stuff came from, so we know where to go when we want to update?

**Comment by beckermr** [comment]
Hold off please on merging. Testing things

**Review by beckermr — CHANGES_REQUESTED** [review]
I need to test some stuff.

**Comment by beckermr** [comment]
ok a few items on my changes here:

1. They come from the global conda-forge pinnings file (https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/conda_build_config.yaml).
2. This configuration is inherently fragile. The compiler versions are zipped with a bunch of other things in the main pinning files. If that zipping gets messed up, rerenders can fail or worse you could get a bad one.
3. Once conda-forge updates in compilers again, stackvana will break for a while as it did now. Not your problem, mostly a note for future me.
4. The correct long-term solution is to make a PR to pybind11 upstream to make its ABI flags more precise. Isuru has a good idea of what is needed and so we can ask him if we want to do that ourselves. 
5. conda-forge itself is going to remove the compiler abi pinning for its packages that use the pybind11-abi flag, but not for normal pybind11-based builds.

**Review by beckermr — APPROVED** [review]

File: recipe/meta.yaml [hunk 1]
```diff
@@ -162,6 +162,8 @@ outputs:
       commands:
         # grep pin-it output for python dependency
         - pin-it rubin-env | grep "python="
+        # ensure binary compatibility; bump major version if not
+        - [ $(echo __GXX_ABI_VERSION | ${GXX} -E - | tail -1) == 1014 ] # [linux]
```
Comment (beckermr) [inline recipe/meta.yaml:166]: you need quotes to make this a string if you want to use it

**Comment by erykoff** [comment]
@ktlim So this is another approach.  I guess the idea is that this will fail if conda forge rerenders into new versions?  What happens if the next version of gcc comes out and something breaks and we need more time to fix it and we also need to patch the env in the meantime?  Or is this sufficiently rare...

**Comment by beckermr** [comment]
> What happens if the next version of gcc comes out and something breaks and we need more time to fix it and we also need to patch the env in the meantime? Or is this sufficiently rare...

Then you can pin the compilers and remove the pin later.

**Comment by ktlim** [comment]
@conda-forge-admin, please rerender

**Comment by ktlim** [comment]
@eli, right, this tests to make sure we aren't building an incompatible environment without being overly prescriptive.  gcc does not change the ABI version with every compiler revision.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -151,6 +152,8 @@ outputs:
       commands:
         # grep pin-it output for python dependency
         - pin-it rubin-env | grep "python="
+        # ensure binary compatibility; bump major version if not
+        - '[ $(echo __GXX_ABI_VERSION | ${GXX} -E - | tail -1) == 1014 ]'  # [linux]
```
Comment (beckermr) [inline recipe/meta.yaml:156]:
```suggestion
        - '[ $(echo __GXX_ABI_VERSION | ${GXX} -E - | tail -1) == 1014 ]'
```

File: recipe/meta.yaml [hunk 3]
```diff
@@ -161,6 +164,8 @@ outputs:
...
```
Comment (beckermr) [inline recipe/meta.yaml:168]:
```suggestion
        - '[ $(echo __GXX_ABI_VERSION | ${GXX} -E - | tail -1) == 1014 ]'
```

**Review by beckermr — COMMENTED** [review]
This should work on both linux and osx now that I think about it. Testing the compiler versions would be better IMHO. These tests don't cover gfortran ABI changes for example.

**Comment by erykoff** [comment]
There's no `$GXX` defined on macos ...

**Comment by beckermr** [comment]
Right @erykoff. I changed my comment to CXX. I get 1002 locally.

**Comment by erykoff** [comment]
Ah yes, `$CXX` works.

**Comment by ktlim** [comment]
Interesting that `__GXX_ABI_VERSION` would be set by a macOS clang compiler.

My argument for checking the ABI version rather than the compiler version is that the latter is overly strict, as the ABI does not change with each compiler update.  In fact, if the ABI version had not happened to change with this update from 9 to 10, I don't think we would have noticed.

**Comment by beckermr** [comment]
That makes sense @ktlim. We do need to check the gfortran SONAME too though. Plus I don't understand if there are other ABI constraints for pure C code.

My point really is that unless we are confident we can properly enumerate all of the ABI constraints, the compiler version being conservative is likely a good thing.

**Comment by ktlim** [comment]
If I do continue to check `__GXX_ABI_VERSION` but also do so on macOS, wouldn't I need to have two different `commands` entries with different selectors, given that the results are different on the two OSes?

Also, I'm not as worried about gfortran ABI compatibility because I don't think we call FORTRAN code from other packages.

**Comment by beckermr** [comment]
yes you'd need two

🤦

**Pull request merged**